File: sdk/java/src/main/java/com/anyvali/schemas/ObjectSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.UnknownKeyMode;
import com.anyvali.ValidationContext;

import java.util.*;

/**
 * Schema for objects/maps with named properties.
 */
public class ObjectSchema extends Schema<Map<String, Object>> {
    private final Map<String, Schema<?>> properties;
    private final Set<String> required;
    private final UnknownKeyMode unknownKeys;
    private final boolean unknownKeysExplicit;

    /**
     * Create an object schema. If required is null, all properties are required by default.
     */
    public ObjectSchema(Map<String, ? extends Schema<?>> properties, Set<String> required,
                        UnknownKeyMode unknownKeys) {
        this(properties, required, unknownKeys, true);
    }

    private ObjectSchema(Map<String, ? extends Schema<?>> properties, Set<String> required,
                         UnknownKeyMode unknownKeys, boolean unknownKeysExplicit) {
        this.properties = new LinkedHashMap<>(properties);
        this.required = required != null ? new LinkedHashSet<>(required) : new LinkedHashSet<>(properties.keySet());
        this.unknownKeys = unknownKeys != null ? unknownKeys : UnknownKeyMode.STRIP;
        this.unknownKeysExplicit = unknownKeysExplicit;
    }

    public ObjectSchema(Map<String, ? extends Schema<?>> properties) {
        this(properties, null, UnknownKeyMode.STRIP, false);
    }

    public ObjectSchema(Map<String, ? extends Schema<?>> properties, Set<String> required) {
        this(properties, required, UnknownKeyMode.STRIP, false);
    }

    private ObjectSchema(ObjectSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
        this.metadata = other.metadata;
        this.properties = new LinkedHashMap<>(other.properties);
        this.required = new LinkedHashSet<>(other.required);
        this.unknownKeys = other.unknownKeys;
        this.unknownKeysExplicit = other.unknownKeysExplicit;
    }

    private UnknownKeyMode effectiveUnknownKeys() {
        return unknownKeysExplicit ? unknownKeys : UnknownKeyMode.STRIP;
    }

    private UnknownKeyMode exportUnknownKeys() {
        return unknownKeysExplicit ? unknownKeys : UnknownKeyMode.STRIP;
    }

    @Override
    @SuppressWarnings("unchecked")
    protected Object validate(Object input, ValidationContext ctx) {
        if (!(input instanceof Map<?, ?> rawMap)) {
            String received = anyvaliTypeName(input);
            ctx.addIssue(IssueCodes.INVALID_TYPE,
                    "Expected object, received " + received, "object", received);
            return null;
        }

        Map<String, Object> map = (Map<String, Object>) rawMap;
        var result = new LinkedHashMap<String, Object>();

        // Validate known properties
        for (var entry : properties.entrySet()) {
            String key = entry.getKey();
            Schema<?> schema = entry.getValue();

            if (map.containsKey(key)) {
                var childCtx = ctx.child(key);
                Object parsed = schema.runPipeline(map.get(key), childCtx);
                result.put(key, parsed);
            } else if (required.contains(key)) {
                if (schema.hasDefaultValue()) {
                    // Apply default via pipeline
                    var childCtx = ctx.child(key);
                    Object parsed = schema.runPipeline(ABSENT, childCtx);
                    result.put(key, parsed);
                } else {
                    var childCtx = ctx.child(key);
                    childCtx.addIssue(IssueCodes.REQUIRED,
                            "Required field '" + key + "' is missing", key, null);
                }
            } else {
                // Optional and not present - check for defaults
                if (schema.hasDefaultValue()) {
                    var childCtx = ctx.child(key);
                    Object parsed = schema.runPipeline(ABSENT, childCtx);
                    result.put(key, parsed);
                }
            }
        }

        // Handle unknown keys
        Set<String> known = properties.keySet();
        for (String key : map.keySet()) {
            if (!known.contains(key)) {
                switch (effectiveUnknownKeys()) {
                    case REJECT -> {
                        var childCtx = ctx.child(key);
                        childCtx.addIssue(IssueCodes.UNKNOWN_KEY,
                                "Unknown key '" + key + "'", null, key);
                    }
                    case STRIP -> {
                        // Just don't include
                    }
                    case ALLOW -> result.put(key, map.get(key));
                }
            }
        }

        return result;
    }

    @Override
    protected Map<String, Object> toNode() {
        var props = new LinkedHashMap<String, Object>();
        for (var entry : properties.entrySet()) {
            props.put(entry.getKey(), entry.getValue().toPortableNode());
        }
        var node = new LinkedHashMap<String, Object>();
        node.put("kind", "object");
        node.put("properties", props);
        node.put("required", new ArrayList<>(required).stream().sorted().toList());
        node.put("unknownKeys", exportUnknownKeys().getValue());
        return addCommonNodeFields(node);
    }

    @Override
    protected Schema<Map<String, Object>> copy() {
        return new ObjectSchema(this);
    }

    public Map<String, Schema<?>> getProperties() {
        return properties;
    }

    public Set<String> getRequired() {
        return required;
    }

    public UnknownKeyMode getUnknownKeys() {
        return effectiveUnknownKeys();
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/ObjectSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

data class ObjectSchema(
    val properties: Map<String, Schema> = emptyMap(),
    val required: Set<String> = emptySet(),
    val unknownKeys: UnknownKeyMode = UnknownKeyMode.REJECT
) : Schema() {
    override val kind: String = "object"

    fun unknownKeys(mode: UnknownKeyMode) = copy(unknownKeys = mode)

    override fun safeParseWithContext(input: Any?, ctx: ValidationContext): ParseResult {
        if (input !is Map<*, *>) {
            return ParseResult.Failure(
                listOf(
                    ValidationIssue(
                        code = IssueCodes.INVALID_TYPE,
                        path = ctx.path,
                        expected = "object",
                        received = getJsonTypeName(input)
                    )
                )
            )
        }

        @Suppress("UNCHECKED_CAST")
        val inputMap = input as Map<String, Any?>
        val issues = mutableListOf<ValidationIssue>()
        val result = mutableMapOf<String, Any?>()

        // Check required fields
        for (reqKey in required) {
            if (!inputMap.containsKey(reqKey)) {
                val propSchema = properties[reqKey]
                // Check for default
                val defaultVal = getDefault(propSchema)
                if (defaultVal != null) {
                    // Will be handled below
                } else {
                    issues.add(
                        ValidationIssue(
                            code = IssueCodes.REQUIRED,
                            path = ctx.path + reqKey,
                            expected = propSchema?.kind ?: "unknown",
                            received = "undefined"
                        )
                    )
                }
            }
        }

        // Validate known properties
        for ((key, schema) in properties) {
            val childCtx = ctx.child(key)

            if (inputMap.containsKey(key)) {
                val childResult = schema.safeParseWithContext(inputMap[key], childCtx)
                when (childResult) {
                    is ParseResult.Success -> result[key] = childResult.value
                    is ParseResult.Failure -> issues.addAll(childResult.issues)
                }
            } else {
                // Try default
                val defaultVal = getDefault(schema)
                if (defaultVal != null) {
                    val dv = defaultVal.second
                    // Validate the default value
                    val childResult = schema.safeParseWithContext(dv, childCtx)
                    when (childResult) {
                        is ParseResult.Success -> result[key] = childResult.value
                        is ParseResult.Failure -> {
                            issues.add(
                                ValidationIssue(
                                    code = IssueCodes.DEFAULT_INVALID,
                                    path = ctx.path + key,
                                    expected = childResult.issues.firstOrNull()?.expected ?: "",
                                    received = dv?.toString() ?: "null"
                                )
                            )
                        }
                    }
                }
                // else: optional field not present, skip
            }
        }

        // Handle unknown keys
        val knownKeys = properties.keys
        for (key in inputMap.keys) {
            if (key !in knownKeys) {
                when (unknownKeys) {
                    UnknownKeyMode.REJECT -> {
                        issues.add(
                            ValidationIssue(
                                code = IssueCodes.UNKNOWN_KEY,
                                path = ctx.path + key,
                                expected = "undefined",
                                received = key
                            )
                        )
                    }
                    UnknownKeyMode.STRIP -> { /* don't include */ }
                    UnknownKeyMode.ALLOW -> result[key] = inputMap[key]
                }
            }
        }

        return if (issues.isEmpty()) ParseResult.Success(result) else ParseResult.Failure(issues)
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        val result = safeParseWithContext(value, ctx)
        return if (result is ParseResult.Failure) result.issues else emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("object"))
        put("properties", JsonObject(properties.mapValues { it.value.exportNode() }))
        put("required", JsonArray(required.sorted().map { JsonPrimitive(it) }))
        put("unknownKeys", JsonPrimitive(unknownKeys.value))
    }

    companion object {
        fun getDefault(schema: Schema?): Pair<String, Any?>? {
            if (schema == null) return null
            return when (schema) {
                is StringSchema -> if (schema.defaultValue !== StringSchema.UNSET) "string" to schema.defaultValue else null
                is IntSchema -> if (schema.defaultValue !== IntSchema.UNSET) "int" to schema.defaultValue else null
                is NumberSchema -> if (schema.defaultValue !== NumberSchema.UNSET) "number" to schema.defaultValue else null
                is BoolSchema -> if (schema.defaultValue !== BoolSchema.UNSET) "bool" to schema.defaultValue else null
                is NullableSchema -> getDefault(schema.inner)
                is OptionalSchema -> getDefault(schema.inner)
                else -> null
            }
        }
    }
}
